src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -156,6 +201,492 @@ MultiRouterPlanCreate(Query *originalQuery, Query *query,
+							errmsg("cannot perform distributed planning for the given "
+								   "modification"),
+							errdetail(
+								"OFFSET clauses are not allowed in INSERT ... SELECT "
+								"queries")
+
+							));
+		}
+
+		if (subquery->windowClause != NULL)
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot perform distributed planning for the given "
+								   "modification"),
+							errdetail(
+								"Window functions are not allowed in INSERT ... SELECT "
+								"queries")));
+		}
+
+		if (subquery->setOperations != NULL)
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot perform distributed planning for the given "
+								   "modification"),
+							errdetail(
+								"Union, Intersect and Except are currently unsupported are not allowed in INSERT ... SELECT "
+								"queries")));
+		}
+	}
+}
+
+
+/*
+ * ErrorIfNotAllParticipatingTablesAreColocated errors out of all tables
+ * referenced in the query are not colocated.
+ */
+static void
+ErrorIfNotAllParticipatingTablesAreColocated(Query *query)
+{
+	List *relationIdList = RelationIdList(query);
+	ListCell *relationIdCell = NULL;
+	uint64 colocationId = INVALID_COLOCATION_ID;
+	bool tablesAreColocated = true;
+
+	foreach(relationIdCell, relationIdList)
+	{
+		Oid relationId = lfirst_oid(relationIdCell);
+		uint64 currentColocationId = TableColocationId(relationId);
+
+		if (currentColocationId == INVALID_COLOCATION_ID)
+		{
+			tablesAreColocated = false;
+			break;
+		}
+
+		/* set for the first table */
+		if (colocationId == INVALID_COLOCATION_ID)
+		{
+			colocationId = currentColocationId;
+		}
+
+		if (colocationId != currentColocationId)
+		{
+			tablesAreColocated = false;
+			break;
+		}
+	}
+
+
+	if (!tablesAreColocated)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("all participating tables should be colocated")));
+	}
+}
+
+
+/*
+ * ErrorIfInsertPartitionColumnDoesNotMatchSelect checks whether the INSERTed table's
+ * partition column value matches with the any of the SELECTed table's partition column.
+ */
+static void
+ErrorIfInsertPartitionColumnDoesNotMatchSelect(Query *query)
+{
+	ListCell *targetEntryCell = NULL;
+	uint32 rangeTableId = 1;
+	RangeTblEntry *insertRte = linitial(query->rtable);
+	RangeTblEntry *subqueryRte = lsecond(query->rtable);
+	Query *subquery = subqueryRte->subquery;
+	Oid insertRelationId = insertRte->relid;
+	Var *insertPartitionColumn = PartitionColumn(insertRelationId, rangeTableId);
+	bool partitionColumnsMatch = false;
+
+	foreach(targetEntryCell, query->targetList)
+	{
+		TargetEntry *targetEntry = (TargetEntry *) lfirst(targetEntryCell);
+
+		if (IsA(targetEntry->expr, Var))
+		{
+			Var *insertVar = (Var *) targetEntry->expr;
+			AttrNumber originalAttrNo = get_attnum(insertRelationId,
+												   targetEntry->resname);
+			TargetEntry *subqeryTargetEntry = NULL;
+
+			if (originalAttrNo != insertPartitionColumn->varattno)
```

Comment:
That seems hugely restrictive. Why do we want to force partition columns have the same ordinal position? Or am I missing something here?  Shouldn't we just check that the column in the SELECT target list that corresponds to the partition column in the INSERT target list is a partition column? Everything else seems to make things more complicated for no benefit?